### PR TITLE
Feature/installation app

### PR DESCRIPTION
**Pull request by tleekkul**
Add `odin.app.installation` view which will support the following API
POST /installation
Headers: `Authorization: Bearer`
Payload: `{"installation_id": ":installation_id"}`

This will register new odin identity with that installation id and user can use the returned JWT to talk with odin.

We have to use POST because we can't detect which user is talking to Odin.

### CODE REVIEW

**Comment by KayEss** [comment]
Which Bearer authorization is used with the submit?

**Comment by tleekkul** [comment]
The application bearer. So we know which application is trying to set the installation.

**Comment by KayEss** [comment]
> The application bearer. So we know which application is trying to set the installation.

Then we need to make it very clear that this is for server-to-server communication only. We don't want game clients to receive the application service JWT

File: Cpp/odin-views/app_secure.cpp [hunk 1]
```diff
@@ -39,7 +39,8 @@ namespace {
             fostlib::pg::connection &cnx,
             fostlib::json jwt_header,
             fostlib::json jwt_body) {
-        const auto app_id = fostlib::coerce<fostlib::string>(jwt_body["iss"]);
+        auto const jwt_iss = fostlib::coerce<fostlib::string>(jwt_body["iss"]);
+        auto const app_id = jwt_iss.substr(odin::c_app_namespace.value().length());
```
Comment (KayEss) [inline Cpp/odin-views/app_secure.cpp]: This code pre-supposes that the prefix exists on the `iss` field. Probably it should check

File: Cpp/odin/app.cpp [hunk 2]
```diff
@@ -51,7 +51,8 @@ fostlib::jwt::mint odin::app::mint_user_jwt(
     const fostlib::string jwt_secret = odin::c_jwt_secret.value() + app_id;
     fostlib::jwt::mint jwt{fostlib::sha256, jwt_secret, std::move(payload)};
     jwt.subject(identity_id);
-    jwt.claim("iss", app_id);
+    const fostlib::string jwt_iss = odin::c_app_namespace.value() + app_id;
+    jwt.claim("iss", fostlib::json{jwt_iss});
```
Comment (KayEss) [inline Cpp/odin/app.cpp]: This code is perfectly fine, but.... I'd move the string into the `json` when setting the claim. Saves a copy (not that it matters).

File: Cpp/odin/odin.cpp [hunk 3]
```diff
@@ -61,6 +62,9 @@ const fostlib::setting<fostlib::string> odin::c_facebook_endpoint(
 const fostlib::setting<fostlib::json> odin::c_google_aud(
         "odin/odin.cpp", "odin", "Google", fostlib::null, true);
 
+const fostlib::setting<fostlib::string> odin::c_app_namespace(
+        "odin/odin.cpp", "odin", "Application namespace", "https://odin.hot-now.com/app/", true);
```
Comment (KayEss) [inline Cpp/odin/odin.cpp]: This needs to be a generic name here, we have to set the real one in our configuration.

File: tests/app-secure.fg [hunk 4]
```diff
@@ -38,13 +38,13 @@ sql.insert odin.app_ledger {
 GET test/app/secure / 401
 
 # Mint App jwt
-set app_jwt (odin.jwt.mint {"sub": "some_user", "iss": "app01"} APP_TOKEN)
+set app_jwt (odin.jwt.mint {"sub": "some_user", "iss": "https://odin.hot-now.com/app/app01"} APP_TOKEN)
```
Comment (KayEss) [inline tests/app-secure.fg]: These have to be the generic name as well

File: tests/app-secure.fg [hunk 5]
```diff
@@ -38,13 +38,13 @@ sql.insert odin.app_ledger {
...
 set-path testserver.headers ["Authorization"] (cat "Bearer " (lookup app_jwt))
 GET test/app/secure / 200
 GET test/app/secure /test 200
 
 # Wrong App jwt should return 401
-set app_jwt (odin.jwt.mint {"sub": "some_user", "iss": "app02"} APP_TOKEN)
+set app_jwt (odin.jwt.mint {"sub": "some_user", "iss": "https://odin.hot-now.com/app/app02"} APP_TOKEN)
```
Comment (KayEss) [inline tests/app-secure.fg]: And here

File: Cpp/odin-views/app_secure.cpp (hunk 1 above)
Comment (tleekkul) [reply Cpp/odin-views/app_secure.cpp]: If the prefix doesn't exist on the `iss` field, what happens? throw exception?

**Pull request merged**